Add initial CI pipeline

TimothyEDawson opened the pull request:
This will add the initial infrastructure for a Github Actions-based CI workflow. The current scope is to build and cache the Loci dependencies, build Loci with different options, and run `make test`. Planned options for the test matrix include:

- Serial build
- Parallel builds with different mesh partitioners (space-filling curve, Parmetis, and PTScotch)
- GNU (gcc) and LLVM (clang) compilers
- Different MPI libraries (Openmpi, mpich, maybe mvapich2)
- Linux (Ubuntu 24.04) and Mac (MacOS 14) operating systems

TimothyEDawson committed:
[#114] Add dependency build-and-cache job.

TimothyEDawson committed:
[#114] Change Spack root location, remove mirror

TimothyEDawson committed:
[#114] Simplify to test Github Packages

TimothyEDawson committed:
[#114] Re-add build cache before push

TimothyEDawson committed:
[#114] Try full build again

TimothyEDawson commented:
Seems to be mostly working! The cached dependencies can be viewed here: https://github.com/edwardaluke/loci/pkgs/container/spack-buildcache

I'll need to tweak some things for MacOS, as it would appear that apple-clang built in to the image is not treated the same by Spack as the llvm clang available in the Ubuntu image. I also found out that you get 1000 minutes / month on the free tier of Github Actions, but the MacOS images are charged at a 10x rate! So I'll focus on nailing down an Ubuntu-only pipeline before I bring MacOS back into it.

EdwardALuke commented:
We may consider moving the repository over to the MsState OTM github space as I think they have a paid for these extra features.  Let me check what is available via the OTM github space.

EdwardALuke commented:
It looks like this is setup to do a weekly check?  Is that the current setup?  Given the limitiations on the free github it would seem potentially a problem to have this run on every pull request.

TimothyEDawson commented:
> It looks like this is setup to do a weekly check? Is that the current setup? Given the limitations on the free github it would seem potentially a problem to have this run on every pull request.

Right now it's on every pull request, but it should complete very quickly as it will obtain all of the pre-built binaries from the cache rather than rebuilding them from scratch. That way it's effectively on-demand, as any newly-requested items - such as pinning a different version - will be the only ones built. I'd still like to change it to something like a monthly check, or ideally only trigger it when the spack.yaml file has been modified.

Also, wow! It took nearly two hours to build everything! That definitely would have been untenable without caching, even if I were to more carefully streamline the build.

TimothyEDawson committed:
[#114] Extract Spack mirrors into its own file.

TimothyEDawson committed:
[#114] Shallow git clone, correct master branch name.

TimothyEDawson committed:
[#114] WIP: Build and test CI workflow.

Utilizes loci_spack repo to control Loci build options and dependency
resolution. Still need to add logic for compiling in serial mode, and
add debug build as an option to loci_spack repo.

TimothyEDawson committed:
[#114] Only allow manual triggering of buildcache job.

TimothyEDawson committed:
[#114] Set Spack directory for the whole CI job.

TimothyEDawson commented:
Looks like the buildcache job still takes about 10 minutes when the cache is available, so for now I'll set it to manual triggering only. But I'll leave in logic for running when the relevant files are modified that I was working on, commented out.

Right now the CI job seems to be failing to pick up and use the `.spack/repos.yaml` file, which seemed to work fine in my local copy. I'll have to come back to this later.

TimothyEDawson committed:
[#114] Add temporary commands for debugging

TimothyEDawson commented:
Very strange... it seems to find the `.spack/mirrors.yaml` file just fine, but it doesn't seem to see the `.spack/repos.yaml` file. It picks it up automatically on my workstation. I'll keep digging into this.

TimothyEDawson commented:
Aha.... Apparently the setup-spack action sets the following variable to true:

- SPACK_DISABLE_LOCAL_CONFIG: Set this environment variable to completely disable both the system and user configuration directories. Spack will then only consider its own defaults and site configuration locations.

That's probably the issue!

TimothyEDawson committed:
[#114] Undo setup-spack setting we don't want.

TimothyEDawson committed:
[#114] More workflow debugging statements.

TimothyEDawson committed:
[#114] Change how Spack configuration is set up.

Because the setup-spack action sets the SPACK_DISABLE_LOCAL_CONFIG
variable in $GITHUB_ENV, we needed to completely rewrite how the
configuration files are found. Instead of relying on setting the
user-scope directory to the .spack subdirectory, we activate the Spack
environment first and add the needed `.yaml` files into the
environment's scope.

TimothyEDawson committed:
[#114] Add loci spec so Spack concretizes correctly.

TimothyEDawson committed:
[#114] Fix spack develop command.

TimothyEDawson commented:
Looks like it died while compiling cgns2ensight, likely some API changed between when that was written and whatever recent version of CGNS was pulled in by Spack (looks like 4.5.0). This is good to know for whenever someone wants to volunteer to fix that. I should be able to pin CGNS to an earlier version that worked (we had 4.4.0 in the submodules), though I'd have to refresh it within the dependency build job. I'll disable it for now and revisit it soon.

Otherwise it looks like everything else (minus MacOS) is fully functional! So now I just need to add in the logic to turn on debug options and to compile a serial Loci and we should be good to go!! One other thing I'll check is to see how to upload the build output file on a build failure, as that would be invaluable when diagnosing the problem.

TimothyEDawson committed:
[#114] Add debug flag, temporarily remove CGNS.

TimothyEDawson committed:
[#114] Remove extra configuration dumps.

TimothyEDawson commented:
Almost! Loci was successfully compiled, and one of the quickTest stages failed with the following message:

```
--------------------------------------------------------------------------
The Open MPI wrapper compiler was unable to find the specified compiler
/usr/bin/clang++-15 in your PATH.

Note that this compiler was either specified at configure time or in
one of several possible environment variables.
--------------------------------------------------------------------------
```

Couple things I realized I should modify:

1. I need to add the full test matrix to the name, so we can see that this error occurred on the parmetis, openmpi, gcc compiler job. Which immediately makes me wonder why it was looking for clang at all! But `/usr/bin/clang++-15` seems like an additional mistake, as it should just be `clang++`.
2. Now that it's getting to the quickTest stage, I should disable the fail-fast option, which cancels all running jobs if one fails. That way they will all run to completion, and we can see if all of them will fail or only some of them.

TimothyEDawson committed:
[#114] Add test matrix details to job name.

Also disable fail-fast, and make sure quickTest is executed from the OBJ
directory.

TimothyEDawson commented:
Still unsure about the "/usr/bin/clang++-15" error, but I'm thinking I might need to regenerate some of the dependencies. I'm going to try reworking the buildcache job, and also print out some checks during the quickTest step such as `mpicc -show` to hopefully figure out where it's going wrong.

EdwardALuke commented:
The quickTest is currently pretty hard to figure out why things are failing.  It isn't quite as good as the chem quickTest where the tests are run and then summarized at the end so that the failing tests are highlighted and the run information can be inspected.  It probably isn't too hard to adapt to this form.  If you need me to work on the quickTest infrastructure let me know.

TimothyEDawson commented:
Looks like we're nearing the usage limit for the month, as https://github.com/EdwardALuke/loci/actions/metrics/usage states 926 minutes used (free tier should be 1000 minutes / month). Assuming it resets on Sept. 1st, I'll need to hold off on running anything more until then. In the meantime I'll look for ways to further streamline the build. E.g. disabling defaults we don't need, such as PETSc defaulting to Hypre support.

Given the cost of the CI runs - the latest test matrix setup, with 12 variations, used 137 minutes - We should also set it to require approval for each run. I believe that's a toggle in the Github settings for this repo, let me know if you can't find it @EdwardALuke. Otherwise each push to a pull request will quickly eat through the minutes we currently have, even if the contributor knows it's not necessary to run yet. In the future, I'll work towards further separating the build + test step into a minimal CI run (e.g. one point from the full test matrix) and a full matrix CI run intended for less frequent checks of the dev branch.

TimothyEDawson commented:
> The quickTest is currently pretty hard to figure out why things are failing. It isn't quite as good as the chem quickTest where the tests are run and then summarized at the end so that the failing tests are highlighted and the run information can be inspected. It probably isn't too hard to adapt to this form. If you need me to work on the quickTest infrastructure let me know.

I appreciate that, and I'll let you know what changes might be beneficial once I've learned a bit more about the Github infrastructure. I know it's common to upload pass/fail and code coverage reports at the end of a job so I'll investigate what sort of format is typically used, and what tools are used for measuring coverage that we might be able to adapt to this (unless you already know how to generate coverage reports). I don't think anything will be needed within the scope of this pull request, though, as the current issue seems to be limited to something wrong with the Spack setup which I need to debug.

TimothyEDawson committed:
[#114] Add debugging statements for mpi commands.

EdwardALuke commented:
Since the cost is potentially prohibitive, maybe we should do a limited test on the pull request, e.g. the most common configuration.  And then do a more comprehensive test on regular intervals, say weekly or biweekly.  That way if there was some edge case problem we could be notified before it gets too hard to figure out which change caused the problem.

TimothyEDawson committed:
[#114] Enable serial build of Loci.

TimothyEDawson committed:
[#114] Extract build and test into reusable job.

TimothyEDawson committed:
[#114] Add single build-test job for CI.

TimothyEDawson committed:
[#114] Upload stage output files as artifact.

TimothyEDawson committed:
[#114] Add Loci config files to upload.

TimothyEDawson committed:
[#114] Fix indentation error in ci-full.yaml.

TimothyEDawson committed:
[#114] Make mpi not required input.

TimothyEDawson commented:
Some progress:

- The default CI now only runs one configuration (Ubuntu, parallel build, parmetis partitioner, and openmpi), and there's a new workflow for running the full matrix. Both require manual deployment.
- A serial build is now included in the full test matrix.
- On failure, the workflow automatically uploads the configuration files from the OBJ directory and the full build log files from Spack. You can see these at the bottom under "Artifacts", e.g. https://github.com/EdwardALuke/loci/actions/runs/17417948578

Now if I could just figure out what's up with the `make test` step, we'll be good to go!

TimothyEDawson committed:
[#114] Run quickTest during build phase.

TimothyEDawson commented:
@EdwardALuke I was able to get it to run `make test` as part of the build step, this is enabled with `spack install --test=root`. Ideally I'd like to separate out the build, test, and install steps, but I think it's fine to leave that as future work.

With the added feature of uploading the build log, it's much easier to see what the issues are! This time, it seems that quickTest failed because it requested 3 processors to run vogcheck and the Gitlab runner did not have that many slots available. Obviously it's important to be able to test these in parallel, so I'll look into what it's possible to do within a single runner instance - I'm pretty sure they have more than one processor available, so it might be some other issue with the way it's being called from Spack (e.g. could `make -j 4 test` have issues?).

TimothyEDawson commented:
Confirmed that the hosted runners indeed have 4 CPUs: https://docs.github.com/en/actions/reference/runners/github-hosted-runners

I also confirmed that `make test` was called with `-j4`, so I probably need to either find a way to disable that option in the Spack call, or just set `MPINUMPROC=1` in the `test.conf` file. I'll try the latter first.

TimothyEDawson committed:
[#114] Set MPINUMPROC=1 in quickTest/test.conf.

TimothyEDawson commented:
@EdwardALuke [Success!](https://github.com/EdwardALuke/loci/actions/runs/17443461193)

This is ready for use, I think we can merge it in as-is and refine it as we go. Some notes:

- Building and testing for a single point in the test matrix took about 30 minutes all told. Whenever I get around to revisiting this, I'll look for ways to streamline it.
- For manually-triggered workflows, the documentation says I should see an option to start it from the Actions tab in the web interface, but I do not see that. Instead I utilized the [GitHub CLI](https://cli.github.com/).
- The syntax to trigger the single-point workflow is: `gh workflow run ci.yaml --ref [branch name] --field pr_number=[associated pull request number]`. For the full test matrix, use `ci-full.yaml` instead.

Also, I currently have it set to only upload the build log (which does include the log from quickTest) on a job failure. Would you like me to set that to always upload, instead? I don't think it would make much difference in terms of computational or storage limits, and might be nice to have available.

EdwardALuke commented:
I think the only time we will want to look at the build log would be on failure.   But this seems to be something we could change if we find that we really want it later.
